File: backend/src/AnimeShop/API/Admin/Employees.hs
{-# LANGUAGE DataKinds #-}
{-# LANGUAGE TypeOperators #-}
module AnimeShop.API.Admin.Employees where

import AnimeShop.Prelude

type EmployeesAPI =
  "employees" :>
  (
    Get '[JSON] [Entity Employee] :<|>
    ReqBody '[JSON] Employee :> Post '[JSON] (Entity Employee) :<|>
    ReqBody '[JSON] EmployeeId :> Delete '[JSON] () :<|>
    ReqBody '[JSON] (Entity Employee) :> Patch '[JSON] ()
  )

employeesServer :: ServerT EmployeesAPI App
employeesServer = getEmployees :<|> postEmployee :<|> deleteEmployee :<|> patchEmployee

getEmployees :: App [Entity Employee]
getEmployees = runDb $ selectList [] []

postEmployee :: Employee -> App (Entity Employee)
postEmployee = runDb . insertEntity

deleteEmployee :: EmployeeId -> App ()
deleteEmployee = runDb . delete

patchEmployee :: Entity Employee -> App ()
patchEmployee entity = runDb $ replace (entityKey entity) (entityVal entity)

File: frontend/src/reducers/admin/employees.js
import { ADD_EMPLOYEE, GET_EMPLOYEES, DELETE_EMPLOYEE } from '../../actions/types';

export function employees(state = [], action) {
  switch (action.type) {
    case ADD_EMPLOYEE:
      return [ ...state, action.employee];
    case GET_EMPLOYEES:
      return action.employees;
    case DELETE_EMPLOYEE:
      return state.filter(employee => employee.id !== action.id);
    default:
      return state;
  }
}
